blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,376 @@
+# Nim-BLSCurve
+# Copyright (c) 2018-Present Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import
+  ./bls_backend, ./bls_sig_min_pubkey,
+  ./openmp
+
+# BLS Batch Verifier
+# ----------------------------------------------------------------------
+# We use OpenMP here but might want to use a simple threadpool
+# for portability as Mac compiler doesn't ship with OpenMP
+# and while MSVC has OpenMP, it's unsure about Mingw
+#
+# Also as Nim supports for view types and openarray in object fields is experimental
+# we collect/copy the inputs in the object for now by copy.
+# instead of accumulating them in a pairing context.
+# Note that a pairing context is 3+MB
+# while we have
+# - publickey = 48B
+# - signature = 96B
+# - message = 32B (assuming sha256 hashed)
+# hence 176B only
+#
+# TODO: Once ContextMultiAggregateVerify is implemented
+# for Milagro/Miracl, this wouldn't need to be in the BLST specific file
+
+type
+  SignatureSet[HashLen: static int] = object
+    ## A (Public Key, Message, Signature) triplet
+    ## that will be batch verified.
+    ## This should not hold GC-ed memory
+    ## as this would complexify multithreading.
+    ## Consequently this assumes that message
+    ## is the output of a fixed size hash function.
+    signature: Signature
+    pubkey: PublicKey
+    message: array[HashLen, byte]
+
+  BatchedBLSVerifier*[HashLen: static int] = object
+    ## A type to batch BLS multi signatures (aggregated or individual)
+    ## verification using multiple cores if compiled with OpenMP
+    sets: seq[SignatureSet[HashLen]]
+
+    # Per-batch contexts for multithreaded batch verification
+    batchContexts: seq[ContextMultiAggregateVerify[DST]]
+
+func init*(T: type BatchedBLSVerifier): T {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  # Impl: A BatchedBLSVerifier still MUST be zero-init
+  # otherwise the sequences fields
+  # like capacity and reserve memory will be wrong.
+  result.sets.setLen(0)
+  result.batchContexts.setLen(0)
+
+func clear*(batcher: var BatchedBLSVerifier) {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  batcher.sets.setLen(0)
+  batcher.batchContexts.setLen(0)
+
+func incl*[HashLen: static int](
+       batcher: var BatchedBLSVerifier[HashLen],
+       public_key: PublicKey,
+       message: array[HashLen, byte],
+       signature: Signature
+     ): bool {.inline.} =
+  ## Include a (public key, message, signature) triplet
+  ## to the batch for verification.
+  ##
+  ## Always return true
```

Comment:
so why not remove the return value and avoid the confusion?

Reply:
Consistency with all the other public API that have an update step like https://github.com/status-im/nim-blscurve/blob/31ada540789fc0081b4fcfd6f693410683235fd7/blscurve/blst/blst_min_pubkey_sig_core.nim#L298-L309

Also even for aggregate which AFAIK cannot fail we've been ask to have a error return value, https://github.com/status-im/nim-blscurve/issues/77

If we add extra sanity checks in there the public API won't need to change, but I can change the comment so that it says "currently always return true but may sanity checks may move at this step"

Reply:
Apparently OpenSSL and BoringSSL also have proc that always return 1: https://github.com/google/boringssl/blob/b117a3a/include/openssl/sha.h#L157-L167

Note sure if there is a deeper reason like dealing with interrupts in embedded devices. Or if it's a requirement in certain environment like MISRA-C.